### PR TITLE
Optimize NOT IN filter for strings

### PR DESCRIPTION
**Pull request by kewang1024**
== NO RELEASE NOTES ==

### CODE REVIEW

**Comment by kewang1024** [comment]
The benchmark is mainly to prove that in situation when we only have one element, we won't have performance regression 

Benchmark                                Mode  Cnt   Score    Error  Units
BenchmarkBytesValues.lookupInExclusive   avgt   20  ≈ 10⁻⁵           ms/op
BenchmarkBytesValues.lookupInMultiRange  avgt   20   0.099 ±  0.006  ms/op

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java [hunk 1]
```diff
@@ -1209,6 +1219,66 @@ public String toString()
...
+
+        public static BytesValuesExclusive of(byte[][] values, boolean nullAllowed)
+        {
+            return new BytesValuesExclusive(values, nullAllowed);
+        }
+
+        @Override
+        public boolean testBytes(byte[] value, int offset, int length)
+        {
+            return !delegate.testBytes(value, offset, length);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java]: check length first:

```
if (!delegate.testLength(length)) {
   return true;
}
```

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java [hunk 2]
```diff
@@ -129,9 +131,18 @@ public static TupleDomainFilter toFilter(Domain domain)
                                 .toArray(byte[][]::new),
                         nullAllowed);
             }
+
+            if (isNotIn(ranges)) {
+                return BytesValuesExclusive.of(
+                        bytesRanges.stream()
+                                .map(BytesRange::getUpper)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java]: use existing `BytesRange::getLower` method here and remove newly added `getUpper`

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java [hunk 3]
```diff
@@ -142,7 +153,7 @@ public static TupleDomainFilter toFilter(Domain domain)
      *
      * The logic is to return true if ranges are next to each other, but don't include the touch value.
      */
-    private static boolean isNanAllowed(List<Range> ranges)
+    private static boolean isNotIn(List<Range> ranges)
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java]: method comment needs to be updated and discussion of NaNs needs to be moved to the caller; I'd move changes to this method into a separate commit 

```
    /**
     * Returns true is ranges represent != or NOT IN filter for double, float or string column.
     *
     * The logic is to return true if ranges are next to each other, but don't include the touch value.
     */
    private static boolean isNotIn(List<Range> ranges)
```

and, in the caller:

```

        if (firstRangeFilter instanceof DoubleRange || firstRangeFilter instanceof FloatRange) {
            // != and NOT IN filters should return true when applied to NaN
            // E.g. NaN != 1.0 as well as NaN NOT IN (1.0, 2.5, 3.6) should return true; otherwise false.
            boolean nanAllowed = isNotIn(ranges);
            return MultiRange.of(rangeFilters, nullAllowed, nanAllowed);
        }

        return MultiRange.of(rangeFilters, nullAllowed, false);
```

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java [hunk 4]
```diff
@@ -1098,7 +1108,7 @@ private BytesValues(byte[][] values, boolean nullAllowed)
             super(true, nullAllowed);
 
             requireNonNull(values, "values is null");
-            checkArgument(values.length > 1, "values must contain at least 2 entries");
+            checkArgument(values.length > 0, "values must contain at least 1 entries");
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java]: message needs updating: `checkArgument(values.length > 0, "values must not be empty");`

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBytesValues.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+@Fork(2)
+@Warmup(iterations = 10, time = 1000, timeUnit = MILLISECONDS)
+@Measurement(iterations = 10, time = 1000, timeUnit = MILLISECONDS)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkBytesValues
+{
+    private final byte[] word = new byte[]{'h', 'h', 'h'};
+
+    @Benchmark
+    public int lookupInExclusive(BenchmarkBytesValues.BenchmarkData data)
```
Comment (mbasmanova) [inline presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBytesValues.java]: I don't think this benchmark is valid. In general you are not supposed to perform same operation in a loop like this, rather let JMH do that. Re-writing the benchmark to remove the loop shows same performance for these two operations:

```
    @Benchmark
    public boolean lookupInExclusive(BenchmarkBytesValues.BenchmarkData data)
    {
        return data.exclusiveFilter.testBytes(word, 0, 3);
    }

    @Benchmark
    public boolean lookupInMultiRange(BenchmarkBytesValues.BenchmarkData data)
    {
        return data.multiRangeFilter.testBytes(word, 0, 3);
    }

# Run complete. Total time: 00:01:23

Benchmark                                Mode  Cnt   Score    Error  Units
BenchmarkBytesValues.lookupInExclusive   avgt   20  ≈ 10⁻⁵           ms/op
BenchmarkBytesValues.lookupInMultiRange  avgt   20  ≈ 10⁻⁵           ms/op
```

Writing good benchmark is not a requirement for the PR. If you want to write it, I suggest to start by writing benchmark for the main use case, e.g. in list with multiple entries, then see when performance difference is for very large in lists (e.g. hundreds of values), then make sure that in-list with one entry or != filter doesn't regress. Here is where I would start:

```
@OutputTimeUnit(NANOSECONDS)

    // NOT IN ("apple", "grape", "orange")
    // "abc", "apple", "banana", "grape", "orange", "peach"
    private final byte[][] testWords = new byte[][] {"abc".getBytes(), "apple".getBytes(), "banana".getBytes(), "grape".getBytes(), "orange".getBytes(), "peach".getBytes()};

    @Benchmark
    public int lookupInExclusive(BenchmarkBytesValues.BenchmarkData data)
    {
        int hits = 0;
        for (int i = 0; i < testWords.length; i++) {
            if (data.exclusiveFilter.testBytes(testWords[i], 0, testWords[i].length)) {
                hits++;
            }
        }

        return hits;
    }

    @Benchmark
    public int lookupInMultiRange(BenchmarkBytesValues.BenchmarkData data)
    {
        int hits = 0;
        for (int i = 0; i < testWords.length; i++) {
            if (data.multiRangeFilter.testBytes(testWords[i], 0, testWords[i].length)) {
                hits++;
            }
        }

        return hits;
    }

    @State(Scope.Thread)
    public static class BenchmarkData
    {
        private TupleDomainFilter exclusiveFilter;
        private TupleDomainFilter multiRangeFilter;

        @Setup
        public void setup()
                throws Exception
        {
            exclusiveFilter = BytesValuesExclusive.of(new byte[][] {"apple".getBytes(), "grape".getBytes(), "orange".getBytes()}, false);
            multiRangeFilter = TupleDomainFilter.MultiRange.of(
                    ImmutableList.of(
                            BytesRange.of(null, true, "apple".getBytes(), true, false),
                            BytesRange.of("apple".getBytes(), true, "grape".getBytes(), true, false),
                            BytesRange.of("grape".getBytes(), true, "orange".getBytes(), true, false),
                            BytesRange.of("orange".getBytes(), true, null, true, false)),
                    false,
                    false);
        }
    }

Benchmark                                Mode  Cnt   Score   Error  Units
BenchmarkBytesValues.lookupInExclusive   avgt   20  63.112 ± 0.811  ns/op
BenchmarkBytesValues.lookupInMultiRange  avgt   20  93.394 ± 0.530  ns/op
```

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java (hunk 1 above)
Comment (kewang1024) [reply presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilter.java]: Sorry, got it lost during rebase, will add it back

File: presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBytesValues.java (hunk 5 above)
Comment (kewang1024) [reply presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBytesValues.java]: Thank you for the detailed suggestion, it's really helpful!

File: presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java [hunk 6]
```diff
@@ -131,18 +131,22 @@ public static TupleDomainFilter toFilter(Domain domain)
...
+
+        return MultiRange.of(rangeFilters, nullAllowed, false);
     }
 
     /**
-     * Returns true is ranges represent != or NOT IN filter. These types of filters should
-     * return true when applied to NaN.
-     *
-     * E.g. NaN != 1.0 as well as NaN NOT IN (1.0, 2.5, 3.6) should return true; otherwise false.
+     * Returns true is ranges represent != or NOT IN filter for double, float or string column.
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java]: @kewang1024 string columns are not yet supported here; either add support or update comment

Comment (kewang1024) [reply presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java]: Nice catch! Thanks

**Comment by kewang1024** [comment]
> @kewang1024 Looks good. Would you include benchmark results into commit message?

Sure, will do!

**Comment by kewang1024** [comment]
# Run complete. Total time: 00:01:23

Benchmark                                Mode  Cnt    Score   Error  Units
BenchmarkBytesValues.lookupInExclusive   avgt   20   68.627 ± 1.291  ns/op
BenchmarkBytesValues.lookupInMultiRange  avgt   20  100.337 ± 2.642  ns/op

**Comment by mbasmanova** [comment]
@kewang1024 It would be interesting to benchmark IN filter with large number of values. Let's say 1000. One way to do that would be to generate 1000 random numbers  in [0, 10_000) range and convert them to strings - that would be the IN list. Then, in the benchmark method, generate a a random number in [0, 10_000) range, convert to string and test. (I don't know though if number generation and converting to string would take signifiant time by themselves. One could check by having another benchmark that did just that.)

**Comment by kewang1024** [comment]
> @kewang1024 It would be interesting to benchmark IN filter with large number of values. Let's say 1000. One way to do that would be to generate 1000 random numbers in [0, 10_000) range and convert them to strings - that would be the IN list. Then, in the benchmark method, generate a a random number in [0, 10_000) range, convert to string and test. (I don't know though if number generation and converting to string would take signifiant time by themselves. One could check by having another benchmark that did just that.)

Did you mean NOT IN filter or IN filter?

Also do we write this benchmark in current `BenchmarkBytesValues` as well?

**Comment by mbasmanova** [comment]
> Did you mean NOT IN filter or IN filter?

NOT IN. Although, once you have a benchmark for NOT IN, it would be trivial to add IN filter there. That would also tell us if there is any performance difference between IN and NOT IN. I expect no measurable difference, but would be good to confirm.

> Also do we write this benchmark in current BenchmarkBytesValues as well?

Yes. It can be a follow up PR though.
